Allow checks config to fail independently of logs config

achntrl opened the pull request:
### What does this PR do?

Allows the logs config to be sent if the checks config parsing failed in AD. 
I refactor the `extractTemplatesFromMap` function to return an array of errors. I add the info whether checks or/and logs parsing failed.

I still handle the errors in the `providers` to give context on the errors (did it happen in a pod / docker container / ecs ?)

### Motivation

In Autodiscovery, when the parsing of checks config fails, the logs configs is not parsed and fails. We want to be able to send logs even if the checks config is broken

### Additional Notes

Ideally we would display something in the `agent status` (or `agent configcheck`?) when we fail to parse a config. I'm not sure there is a simple way of doing this for the logs.

achntrl committed:
Allow checks config to fail independently of logs config

achntrl committed:
Add a test case for the check fail + log succeed

achntrl committed:
Refactor extractTemplatesFromMap

xvello commented on pkg/autodiscovery/providers/docker.go:
```diff
@@ -120,16 +120,18 @@ func (d *DockerConfigProvider) IsUpToDate() (bool, error) {
-		case len(c) == 0:
+		c, errors := extractTemplatesFromMap(docker.ContainerIDToEntityName(cID), labels, dockerADLabelPrefix)
+
+		if len(c) == 0 {
```
If `extractTemplatesFromMap` returns no config but errors, will these be logged if we continue here?

achntrl replied:
Yes thanks, that's a bug. 
That's not necessary because to check length and `continue` since append a nil no-op

xvello commented on pkg/autodiscovery/providers/utils.go:
```diff
@@ -100,22 +100,23 @@ func buildTemplates(key string, checkNames []string, initConfigs, instances []in
-func extractTemplatesFromMap(key string, input map[string]string, prefix string) ([]integration.Config, error) {
+func extractTemplatesFromMap(key string, input map[string]string, prefix string) ([]integration.Config, []error) {
 	configs := make([]integration.Config, 0)
+	errors := make([]error, 0)
```
nit: to avoid useless allocations, we can make it `var errors []error`. Same for configs.
[`nil` is a valid default value for go slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices)

xvello commented on pkg/autodiscovery/providers/docker.go:
```diff
@@ -120,16 +120,18 @@ func (d *DockerConfigProvider) IsUpToDate() (bool, error) {
-		default:
-			configs = append(configs, c...)
 		}
+		if len(errors) > 0 {
```
nit: ranging on a `nil` slice is a no-op, so you can skip the length test and directly iterate

xvello reviewed (changes requested):
A few nits on the style, but I have a concern about errors really going though to the logs if `extractTemplatesFromMap` returns `nil, []errors{.......}`

xvello commented:
I removed the `noreno` tag as this is a user-facing change from 6.5. It requires a `fixes` item

achntrl committed:
Fix bug in providers

The error message would be ignored if `extractTemplatesFromMap`
returned an array of config of len == 0

achntrl committed:
Add reno

xvello reviewed (approved):
LGTM 👍

ajacquemot commented on pkg/autodiscovery/providers/kubelet.go:
```diff
@@ -90,18 +90,14 @@ func parseKubeletPodlist(podlist []*kubelet.Pod) ([]integration.Config, error) {
-				configs = append(configs, c...)
 
+			for _, err := range errors {
+				log.Errorf("Can't parse template for pod %s: %s", pod.Metadata.Name, err)
```
Popping in the discussion a bit late but I would have accumulated the messages into a single error instead.
